flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/DefaultLeaderElectionService.java

```diff
@@ -100,6 +100,7 @@ public final void stop() throws Exception {
             if (!running) {
                 return;
             }
+            leaderContender.revokeLeadership();
```

Comment:
@wangyang0918 Was there a rationale to omit the revoking of the leadership when stopping the service? We even added a dedicated test for that in `DefaultLeaderElectionServiceTest`. :thinking: 

To help gettin back the memory: The commit belongs to PR #13644 (FLINK-19542).

Reply:
Ok, digging into a the PR a bit more shows that the old [ZooKeeperLeaderElectionService](https://github.com/apache/flink/pull/13644/files?show-deleted-files=true&show-viewed-files=true&file-filters%5B%5D=#diff-de2fa2be46667dad62b57be6fefa6135043e19888afd453b634792241785c8e0L147) didn't call the revoke. So, I guess, that's just keeping the implementation like that.

But just in general from a theoretical standpoint: We do want to consistently call revoke during shutdown to trigger the clean up through revokeLeadership to stick to the right protocol. That, as a consequence, would make cleanup in the `LeaderContender.stop()` implementation obsolete. WDYT? :thinking:

Reply:
From what I've seen within the code, any implementation does essentially call redundant code cleaning up artifacts in the classes corresponding close implementation and its `revokeLeadership()`.

* `DefaultDispatcherRunner`: `revokeLeadership()` ([DefaultDispatcherRunner:189ff](https://github.com/apache/flink/blob/2d1d61c68d81eba0f721b10eaf2a7246aa0814c2/flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/DefaultDispatcherRunner.java#L189)) only calls `stopDispatcherLeaderProcess` which is also called as part of the `DefaultDispatcherRunner.closeAsync` call in [DefaultDispatcherRunner:96](https://github.com/apache/flink/blob/2d1d61c68d81eba0f721b10eaf2a7246aa0814c2/flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/runner/DefaultDispatcherRunner.java#L96).
* `WebMonitorEndpoint` doesn't do anything except for logging the leadership loss ([WebMonitorEndpoint:1101](https://github.com/apache/flink/blob/be240dea803fc82299c4711ebe656c1c4b159ca9/flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/WebMonitorEndpoint.java#L1101)
* `ResourceManagerServiceImpl`: `revokeLeadership()` ([ResourceManagerServiceImpl:214ff](https://github.com/apache/flink/blob/26aa543b3bbe2b606bbc6d332a2ef7c5b46d25eb/flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ResourceManagerServiceImpl.java#L214)) calls `stopLeaderResourceManager()` besides some logging which is also called by [ResourceManagerServiceImpl.closeAsync](https://github.com/apache/flink/blob/26aa543b3bbe2b606bbc6d332a2ef7c5b46d25eb/flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ResourceManagerServiceImpl.java#L170)
* `JobMasterServiceLeadershipRunner`: `revokeLeadership` ([JobMasterServiceLeadershipRunner:381ff](https://github.com/apache/flink/blob/d745f5b3f7a64445854c735668afa9b72edb3fee/flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobMasterServiceLeadershipRunner.java#L381)) calls `stopJobMasterServiceProcessAsync` which calls `stopJobMasterService` (see [JobMasterServiceLeadershipRunner:388ff](https://github.com/apache/flink/blob/d745f5b3f7a64445854c735668afa9b72edb3fee/flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobMasterServiceLeadershipRunner.java#L401)) which completes the `jobMasterGatewayFuture` exceptionally and calls `closeAsync` on `jobMasterServiceProcess`. The same things are called in `JobMasterServiceLeadershipRunner.closeAsync` (see [JobMasterServiceLeadershipRunner:126ff](https://github.com/apache/flink/blob/d745f5b3f7a64445854c735668afa9b72edb3fee/flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobMasterServiceLeadershipRunner.java#L126)).

That keeps me thinking that there's actually a clearer process possible from a logical standpoint: A `LeaderContender` needs to call `revokeLeadership` before continuing with its shutdown.

Reply:
Sorry for the late response.

I believe that you are right we have some redundant codes in `close/closeAsync()` and `revokeLeadership()` of all `LeaderContender` implementations. And I get your point why you want to call the `revokeLeadership` in the shutdown process.

However, I strongly have a mind that the leader information stored in the ZNode and ConfigMap needs to be cleaned up if a leader contender is revoked leadership. Before this change, this is true since the `revokeLeadership` happens along with `clearConfirmedLeaderInformation`.

Moreover, from the point of view of leader elector, I think it does not lose leadership when stopping since the timeout has not yet expired.

So I lean to not explicitly call the `leaderContender.revokeLeadership()` here.

Reply:
Thanks for clarifying things in this matter.
> However, I strongly have a mind that the leader information stored in the ZNode and ConfigMap needs to be cleaned up if a leader contender is revoked leadership. Before this change, this is true since the revokeLeadership happens along with clearConfirmedLeaderInformation.

This could be fixed by adding to the contract of `DefaultLeaderElectionService` that it calls `onRevokeLeadership()` when before stopping the service. Essentially, leadership can be revoked by the HA backend or by itself when calling `stop`. The leader elector still having leadership shouldn't be a problem because only the contender deregisters himself from the leadership process.

Reply:
> Essentially, leadership can be revoked by the HA backend or by itself when calling `stop`. 

This make sense to me. Then I agree to call the `onRevokeLeadership()` as part of the `DefaultLeaderElectionService.stop()`. It means the leader information stored in the ZNode/ConfigMap will be cleared when the contender is stopping. This should be the expected behavior.

Reply:
Thanks for clarification

Reply:
I re-iterated over that issue once more and realized that I looked at it from the wrong ankle: There's no reason to call `onRevokeLeadership` as part of the `LeaderElectionService.stop()` call. `stop()` is always called by the contender as part of its shutdown process. Therefore, the contender is already aware of shutting down the leader election process. That's why, we can consider the `onRevokeLeadership` call as obsolete. The `onRevokeLeadership` call would be ignored, anyway, because of the `running` flag that is set to `false` in any `LeaderContender` implementation before shutting down the `LeaderElectionService`. That flag is checked in the `onRevokeLeadership` implementations and would omit any logic. Working around this would require a bigger refactoring and is not necessary.

Therefore, instead of adding the `onRevokeLeadership` to the shutdown procedure we can remove this call from the `LeaderElectionService` implementations that do call it (also including the `MultipleComponentLeaderElectionService` which does call it without any necessity).

I'm gonna update the PR to consider this.